utils/sync/wait_with_timeout.go

```diff
@@ -0,0 +1,22 @@
+package sync
+
+import (
+	"sync"
+	"time"
+)
+
+// WaitTimeout waits for the waitgroup for the specified max timeout.
+// Returns true if waiting timed out.
+func WaitTimeout(wg *sync.WaitGroup, timeout time.Duration) bool {
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
+		wg.Wait()
+	}()
+	select {
+	case <-c:
+		return false // completed normally
+	case <-time.After(timeout):
+		return true // timed out
+	}
```

Comment:
There might be cases where the routine never returns, why not just using normal channels and/or contexts?
Also considering the package where this lives people might be keen to use it in production code as well, not just tests, and this should be avoided imho.

Reply:
This morning, after some consideration followed by reviewing this PR yesterday, I wrote this implementation: https://github.com/rudderlabs/rudder-server/pull/1845

It's an implementation that shouldn't leak, let me know your opinion. If you like the package I can remove it from draft and have it reviewed.

Reply:
I closed the PR and I vote for rewriting the whole `uploader` given that it is a small package it looks very doable :+1: